### PR TITLE
perf: improved performance of comparison (equals) between not compatible types and nulls

### PR DESCRIPTION
**Pull request by lvca**
This fix dramatically improves the performance of traversing when different types are encountered or simply null values. This piece of code should be designed to be fast because invoked millions of times with queries. It is a bad practice to throw Java exceptions instead of simply returning false.

It's already patched in ArcadeDB codebase, but I think it should be fixed in the main codebase (especially because GremlinValueComparator is not pluggable)

### CODE REVIEW

**Comment by mikepersonick** [comment]
This change looks perfectly reasonable to me. +1

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/GremlinValueComparator.java [hunk 1]
```diff
@@ -129,8 +129,18 @@ public boolean equals(final Object f, final Object s) {
...
+            if (eitherAreNaN(f, s))
+                return false;
+
+            // For Compare we do not cross type boundaries, including null
+            if (!comparable(f, s))
+                return false;
+
             try {
-                return this.compare(f, s) == 0;
+              // comparable(f, s) assures that type(f) == type(s)
```
[the remark this answers is not in the document]
Comment (lvca) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/GremlinValueComparator.java:141]: If you remove the try/catch block some tests won't pass

**Comment by xiazcy** [comment]
VOTE +1

**Comment by Cole-Greer** [comment]
Just wanted to chime in really quick before it gets merged and ask if it's worth pushing this into 3.6-dev? Looks like a non breaking change which would help 3.6 users

**Comment by xiazcy** [comment]
Cherry picked to https://github.com/apache/tinkerpop/commit/44c888dec68d66e56fd2acc8dbf09d7d1eed2aa5 and added changelog entry.

**Pull request closed without merging**